packages/block-editor/src/components/contrast-checker/index.js

```diff
@@ -64,61 +71,166 @@ function ContrastChecker( {
 		backgroundColor || fallbackBackgroundColor
 	);
 	const colordTextColor = colord( textColor || fallbackTextColor );
-	const textColorHasTransparency = colordTextColor.alpha() < 1;
+	const colordLinkColor = colord( linkColor || fallbackLinkColor );
+	const textColorHasTransparency =
+		colordTextColor && colordTextColor.alpha() < 1;
+	const linkColorHasTransparency =
+		colordLinkColor && colordLinkColor.alpha() < 1;
 	const backgroundColorHasTransparency = colordBackgroundColor.alpha() < 1;
+
 	const hasTransparency =
-		textColorHasTransparency || backgroundColorHasTransparency;
-	const isReadable = colordTextColor.isReadable( colordBackgroundColor, {
-		level: 'AA',
-		size:
-			isLargeText || ( isLargeText !== false && fontSize >= 24 )
-				? 'large'
-				: 'small',
-	} );
+		backgroundColorHasTransparency ||
+		textColorHasTransparency ||
+		linkColorHasTransparency;
+
+	const textSize =
+		isLargeText || ( isLargeText !== false && fontSize >= 24 )
+			? 'large'
+			: 'small';
+
+	const isTextColorReadable =
+		hasTextColor &&
+		colordTextColor.isReadable( colordBackgroundColor, {
+			level: 'AA',
+			size: textSize,
+		} );
+
+	const isLinkColorReadable =
+		hasLinkColor &&
+		colordLinkColor.isReadable( colordBackgroundColor, {
+			level: 'AA',
+			size: textSize,
+		} );
 
 	// Don't show the message if the text is readable AND there's no transparency.
 	// This is the default.
-	if ( isReadable && ! hasTransparency ) {
-		return null;
+	if ( ! hasTransparency ) {
+		if ( ! hasLinkColor && isTextColorReadable ) {
+			return null;
+		}
+
+		if ( ! hasTextColor && isLinkColorReadable ) {
+			return null;
+		}
+
+		if ( isTextColorReadable && isLinkColorReadable ) {
+			return null;
+		}
 	}
 
 	if ( hasTransparency ) {
+		// If there's transparency, don't show the message if the alpha checker is disabled.
+		if ( ! enableAlphaChecker ) {
+			return null;
+		}
+
+		// If the background has transparency, don't show any warnings.
+		if (
+			backgroundColorHasTransparency &&
+			( ! textColorHasTransparency || ! linkColorHasTransparency )
+		) {
+			return null;
+		}
+
+		// Only text color.
+		if (
+			! hasLinkColor &&
+			isTextColorReadable &&
+			! textColorHasTransparency
+		) {
+			return null;
+		}
+
+		if (
+			! hasTextColor &&
+			isLinkColorReadable &&
+			! linkColorHasTransparency
+		) {
+			return null;
+		}
+
 		if (
-			// If there's transparency, don't show the message if the alpha checker is disabled.
-			! enableAlphaChecker ||
-			// If the alpha checker is enabled, we only show the warning if the text has transparency.
-			( isReadable && ! textColorHasTransparency )
+			isTextColorReadable &&
+			! textColorHasTransparency &&
+			isLinkColorReadable &&
+			! linkColorHasTransparency
 		) {
 			return null;
 		}
 	}
 
+	// Flag to warn about transparency only if the text is otherwise readable according to colord
+	// to ensure the readability warnings take precedence.
+	let shouldShowTransparencyWarning = false;
```

Comment:
Please ignore all these conditions above and below for now. 

They're just there to get the logic right first.

Reply:
It will be a hook eventually.